File: tests/abi/context_menu_layout.cpp
#include "pluginterfaces/vst/ivstcontextmenu.h"

#include <cstddef>
#include <cstdio>

#define PRINT_TYPE(Type) std::printf (#Type " size %zu align %zu\n", sizeof (Steinberg::Vst::Type), alignof (Steinberg::Vst::Type))
#define PRINT_OFFSET(Type, Field) std::printf (#Type "." #Field " offset %zu\n", offsetof (Steinberg::Vst::Type, Field))

template <typename T>
static void print_iid (const char* name, const T& tuid)
{
	std::printf ("%s iid", name);
	for (int index = 0; index < 16; ++index)
		std::printf (" %02X", static_cast<unsigned char> (tuid[index]));
	std::printf ("\n");
}

int main ()
{
	std::printf ("IContextMenuItem.kIsSeparator %d\n", Steinberg::Vst::IContextMenuItem::kIsSeparator);
	std::printf ("IContextMenuItem.kIsDisabled %d\n", Steinberg::Vst::IContextMenuItem::kIsDisabled);
	std::printf ("IContextMenuItem.kIsChecked %d\n", Steinberg::Vst::IContextMenuItem::kIsChecked);
	std::printf ("IContextMenuItem.kIsGroupStart %d\n", Steinberg::Vst::IContextMenuItem::kIsGroupStart);
	std::printf ("IContextMenuItem.kIsGroupEnd %d\n", Steinberg::Vst::IContextMenuItem::kIsGroupEnd);

	PRINT_TYPE (IContextMenuItem);
	PRINT_OFFSET (IContextMenuItem, name);
	PRINT_OFFSET (IContextMenuItem, tag);
	PRINT_OFFSET (IContextMenuItem, flags);

	print_iid ("IComponentHandler3", Steinberg::Vst::IComponentHandler3_iid);
	print_iid ("IContextMenuTarget", Steinberg::Vst::IContextMenuTarget_iid);
	print_iid ("IContextMenu", Steinberg::Vst::IContextMenu_iid);
	return 0;
}

File: tools/context_menu_layout.zig
const std = @import("std");
const context_menu = @import("vst3-zig").pluginterfaces.vst.ivstcontextmenu;

pub fn main() !void {
    const stdout = std.io.getStdOut().writer();
    try stdout.print("IContextMenuItem.kIsSeparator {}\n", .{context_menu.IContextMenuItem.Flags.kIsSeparator});
    try stdout.print("IContextMenuItem.kIsDisabled {}\n", .{context_menu.IContextMenuItem.Flags.kIsDisabled});
    try stdout.print("IContextMenuItem.kIsChecked {}\n", .{context_menu.IContextMenuItem.Flags.kIsChecked});
    try stdout.print("IContextMenuItem.kIsGroupStart {}\n", .{context_menu.IContextMenuItem.Flags.kIsGroupStart});
    try stdout.print("IContextMenuItem.kIsGroupEnd {}\n", .{context_menu.IContextMenuItem.Flags.kIsGroupEnd});

    try printType(stdout, "IContextMenuItem", context_menu.IContextMenuItem);
    try printOffset(stdout, "IContextMenuItem", "name", context_menu.IContextMenuItem, "name");
    try printOffset(stdout, "IContextMenuItem", "tag", context_menu.IContextMenuItem, "tag");
    try printOffset(stdout, "IContextMenuItem", "flags", context_menu.IContextMenuItem, "flags");

    try printType(stdout, "IComponentHandler3", context_menu.IComponentHandler3);
    try printType(stdout, "IContextMenuTarget", context_menu.IContextMenuTarget);
    try printType(stdout, "IContextMenu", context_menu.IContextMenu);
    try printTuid(stdout, "IComponentHandler3", context_menu.icomponent_handler3_iid);
    try printTuid(stdout, "IContextMenuTarget", context_menu.icontext_menu_target_iid);
    try printTuid(stdout, "IContextMenu", context_menu.icontext_menu_iid);
}

fn printType(writer: anytype, comptime name: []const u8, comptime Type: type) !void {
    try writer.print("{s} size {} align {}\n", .{ name, @sizeOf(Type), @alignOf(Type) });
}

fn printOffset(writer: anytype, comptime type_name: []const u8, comptime field_label: []const u8, comptime Type: type, comptime field_name: []const u8) !void {
    try writer.print("{s}.{s} offset {}\n", .{ type_name, field_label, @offsetOf(Type, field_name) });
}

fn printTuid(writer: anytype, comptime name: []const u8, bytes: [16]u8) !void {
    try writer.print("{s} iid", .{name});
    for (bytes) |byte| {
        try writer.print(" {X:0>2}", .{byte});
    }
    try writer.writeByte('\n');
}
